File: scripts/good_first_issues_by_component.R
library(gh)
library(dplyr)
library(purrr)
library(stringr)
library(tibble)

# Fetch all open "good-first-issue" issues in a single paginated call
message("Fetching open good-first-issue issues...")

issues <- gh(
  "/search/issues",
  q = "repo:apache/arrow is:issue state:open label:good-first-issue",
  .limit = Inf
)

message(sprintf("Found %d issues", length(issues$items)))

# Extract component labels from each issue
extract_components <- function(issue) {
  labels <- map_chr(issue$labels, "name")
  components <- labels[str_starts(labels, "Component: ")]
  if (length(components) == 0) {
    return(tibble(component = "(No component)"))
  }
  tibble(component = str_remove(components, "^Component: "))
}

# Build a table of all component assignments
component_data <- map_dfr(issues$items, extract_components)

# Count by component
summary <- component_data |>
  count(component, name = "open_good_first_issues") |>
  arrange(desc(open_good_first_issues))

print(summary, n = Inf)

message(sprintf("\nTotal: %d issues across %d components",
                sum(summary$open_good_first_issues),
                nrow(summary)))

File: scripts/good_first_issues_by_component.py
"""
Fetch and summarize open good-first-issue issues by component.

This is a utility script for quick analysis, not part of the daily update.
"""

import os
import sys
from collections import Counter

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from scripts.github_helpers import gh_search_count, HTTP_HEADERS, OWNER, REPO
import requests


def main():
    print("Fetching open good-first-issue issues...")

    # Fetch all matching issues (paginated)
    issues = []
    page = 1
    while True:
        resp = requests.get(
            "https://api.github.com/search/issues",
            params={
                "q": f"repo:{OWNER}/{REPO} is:issue state:open label:good-first-issue",
                "per_page": 100,
                "page": page,
            },
            headers=HTTP_HEADERS,
        )
        resp.raise_for_status()
        data = resp.json()
        issues.extend(data["items"])

        if len(data["items"]) < 100:
            break
        page += 1

    print(f"Found {len(issues)} issues")

    # Extract component labels
    components = Counter()
    for issue in issues:
        issue_components = [
            label["name"].replace("Component: ", "")
            for label in issue["labels"]
            if label["name"].startswith("Component: ")
        ]
        if not issue_components:
            components["(No component)"] += 1
        else:
            for comp in issue_components:
                components[comp] += 1

    # Print summary
    print("\nComponent                    Count")
    print("-" * 40)
    for comp, count in components.most_common():
        print(f"{comp:<30} {count:>5}")

    print(f"\nTotal: {len(issues)} issues across {len(components)} components")


if __name__ == "__main__":
    main()
